File: python/strings/0005-longest-palindromic-substring.py
def longestPalindrome(s: str) -> str:
    n = len(s)
    res = s[0]
    for i in range(n):
        # Check once for even palindrome size
        mid = i
        left = i - 1
        while mid < n and left >= 0 and s[mid] == s[left]:
            left -= 1
            mid  += 1
        res = max(res, s[left+1:mid], key=lambda x: len(x))
        
        # Check twice for odd palindrome size
        left  = i - 1
        right = i + 1
        while left >= 0 and right < n and s[left] == s[right]:
            left  -= 1
            right += 1
        res = max(res, s[left+1:right], key=lambda x: len(x))
        
    return res

File: cpp/strings/0005-longest-palindromic-substring.cpp
#include <string>

using namespace std;


string longestPalindrome(string s) {
    int start = 0;
    int len = 0;
    for (int i = 0; i < s.length(); i++) {
        int left = i - 1;
        int right = i;
        while (left >= 0 && right < s.length() && s[left] == s[right]) {
            left--;
            right++;
        }
        if (right - (left + 1)> len) {
            start = left + 1;
            len = right - (left + 1);
        }
        
        left = i - 1;
        right = i + 1;
        while (left >= 0 && right < s.length() && s[left] == s[right]) {
            left--;
            right++;
        }
        if (right - (left + 1) > len) {
            start = left + 1;
            len = right - (left + 1);
        }
    }
    return s.substr(start, len);
}
